base_layer/wallet/src/transaction_manager.rs

```diff
@@ -0,0 +1,529 @@
+
+        Ok(SenderMessage::Single(Box::new(msg)))
+    }
+
+    /// Accept the public reply from a recipient and apply the reply to the relevant transaction protocol
+    pub fn accept_recipient_reply(
+        &mut self,
+        recipient_reply: RecipientSignedTransactionData,
+    ) -> Result<(), TransactionManagerError>
+    {
+        let mut marked_for_removal = None;
+
+        for (tx_id, stp) in self.pending_outbound_transactions.iter_mut() {
+            if stp.check_tx_id(recipient_reply.tx_id) && stp.is_collecting_single_signature() {
+                stp.add_single_recipient_info(recipient_reply.clone())?;
+                stp.finalize(KernelFeatures::empty())?;
+                let tx = stp.get_transaction()?;
+                self.completed_transactions.insert(recipient_reply.tx_id, tx.clone());
+                marked_for_removal = Some(tx_id.clone());
+                break;
+            }
+        }
+
+        if marked_for_removal.is_none() {
+            return Err(TransactionManagerError::TransactionDoesNotExistError);
+        }
+
+        if let Some(tx_id) = marked_for_removal {
+            self.pending_outbound_transactions.remove(&tx_id);
+        }
+
+        Ok(())
+    }
+
+    /// Accept a new transaction from a sender by handling a public SenderMessage
+    /// Returns the public reply message to be sent back to the sender.
+    pub fn accept_transaction(
+        &mut self,
+        sender_message: SenderMessage,
+        nonce: SecretKey,
+        spending_key: SecretKey,
+    ) -> Result<RecipientSignedTransactionData, TransactionManagerError>
+    {
+        let rtp = ReceiverTransactionProtocol::new(sender_message, nonce, spending_key, OutputFeatures::empty());
+        let recipient_reply = rtp.get_signed_data()?;
+
+        // Check this is not a repeat message i.e. tx_id doesn't already exist in our pending or completed transactions
+        for (tx_id, _) in self.pending_outbound_transactions.iter() {
+            if tx_id == &recipient_reply.tx_id {
+                return Err(TransactionManagerError::RepeatedMessageError);
+            }
+        }
+        for (tx_id, _) in self.pending_inbound_transactions.iter() {
+            if tx_id == &recipient_reply.tx_id {
+                return Err(TransactionManagerError::RepeatedMessageError);
+            }
+        }
+        for (tx_id, _) in self.completed_transactions.iter() {
+            if tx_id == &recipient_reply.tx_id {
+                return Err(TransactionManagerError::RepeatedMessageError);
+            }
+        }
+
+        // Otherwise add it to our pending transaction list and return reply
+        self.pending_inbound_transactions
+            .insert(recipient_reply.tx_id.clone(), rtp);
+
+        Ok(recipient_reply)
+    }
+
+    /// Return a copy of the completed transactions
+    fn get_completed_transactions(&self) -> HashMap<u64, Transaction> {
+        return self.completed_transactions.clone();
```

Comment:
I'd rather return references here, and below

Reply:
I felt that I didn't want to return a ref because it could mean a client of the manager could mess with the internal state? I know we are not doing anything with these completed transactions yet but if we do?

Reply:
After thinking about it for a bit I couldn't think of a use case for my argument so implemented this suggestion.